src/consistent_request.rs

```diff
@@ -0,0 +1,226 @@
+        // in a valid state at all await points.
+        let mut manager = self.clone();
+        let parts_clone = parts.clone();
+        let body_clone = body.clone();
+        let res_sender_clone = res_tx.clone();
+        tokio::spawn(async move {
+            manager
+                .send_to_l2(parts_clone, body_clone, res_sender_clone)
+                .await
+        })
+        .await??;
+
+        loop {
+            let mut manager = self.clone();
+            let parts_clone = parts.clone();
+            let body_clone = body.clone();
+            match tokio::spawn(
+                async move { manager.send_to_builder(parts_clone, body_clone).await },
+            )
+            .await?
+            {
+                Ok(_) => return Ok(()),
+                Err(_) => {
+                    tokio::time::sleep(Duration::from_millis(200)).await;
+                }
+            }
+        }
+    }
+
+    async fn send_to_l2(
+        &mut self,
+        parts: request::Parts,
+        body: Vec<u8>,
+        res_tx: watch::Sender<Option<Result<(response::Parts, Vec<u8>), BoxError>>>,
+    ) -> eyre::Result<()> {
+        let l2_req = HttpRequest::from_parts(parts.clone(), HttpBody::from(body.clone()));
+        let l2_res = self.l2_client.forward(l2_req, self.method.clone()).await;
+
+        // Return the l2 response asap
+        match l2_res {
+            Ok(t) => {
+                let (parts, body) = t.into_parts();
+                let (body_bytes, _) =
+                    http_helpers::read_body(&parts.headers, body, u32::MAX).await?;
+                res_tx.send(Some(Ok((parts, body_bytes))))?;
+                Ok(())
+            }
+            Err(e) => {
+                let msg = format!("failed to send request to l2: {e}");
+                res_tx.send(Some(Err(e)))?;
+                Err(eyre!(msg))
+            }
+        }
+    }
+
+    async fn send_to_builder(&mut self, parts: request::Parts, body: Vec<u8>) -> eyre::Result<()> {
+        let builder_req = HttpRequest::from_parts(parts.clone(), HttpBody::from(body.clone()));
+        let builder_res = self
+            .builder_client
+            .forward(builder_req, self.method.clone())
+            .await;
+
+        match builder_res {
+            Ok(_) => {
+                if self
+                    .has_disabled_execution_mode
+                    .load(std::sync::atomic::Ordering::SeqCst)
+                {
+                    let mut execution_mode = self.execution_mode.lock();
+                    *execution_mode = ExecutionMode::Enabled;
+                    // Drop before aquiring health lock
+                    drop(execution_mode);
+                    self.has_disabled_execution_mode
+                        .store(false, std::sync::atomic::Ordering::SeqCst);
+                    info!(target: "proxy::call", message = "setting execution mode to Enabled");
```

Comment:
should `self.probes.set_health` be reset here if the call failed previously

Reply:
The health should be automatically reset when the builder successfully returns the next payload.